scanpy/experimental/pp/_highly_variable_genes.py

```diff
@@ -27,6 +28,86 @@
 )
 
 
+@nb.njit(parallel=True)
+def calculate_res_sparse(
+    indptr,
+    index,
+    data,
+    sums_genes,
+    sums_cells,
+    residuals,
+    sum_total,
+    clip,
+    theta,
+    n_genes,
+    n_cells,
+):
+    for gene in nb.prange(n_genes):
+        start_idx = indptr[gene]
+        stop_idx = indptr[gene + 1]
+
+        sparse_idx = start_idx
+        var_sum = np.float64(0.0)
+        sum_clipped_res = np.float64(0.0)
+        for cell in range(n_cells):
+            mu = sums_genes[gene] * sums_cells[cell] / sum_total
+            value = np.float64(0.0)
+            if sparse_idx < stop_idx and index[sparse_idx] == cell:
+                value = data[sparse_idx]
+                sparse_idx += 1
+            mu_sum = value - mu
+            pre_res = mu_sum / sqrt(mu + mu * mu / theta)
+            clipped_res = min(max(pre_res, -clip), clip)
```

Comment:
Ok so I don't know if you can do this more elegant. Since genes cell combination also have a residue when the count is 0. We still have to acount for them. I came up with this logic to run over all combination and check if this combination has a count.